### ISSUE TITLE
Remove weight connections from pretrained LSTM

### ISSUE THREAD

**Issue by shaifugpt** [fchollet/keras#7532]
How can we remove some weight connections from pretrained LSTM and retrain it

**Comment by mahnerak** [comment]
The question is not clear.
Do you mean resetting LSTM layer weight? Or restoring network structure from saved model?
Or something else?

**Comment by shaifugpt** [comment]
I want to set that weight to 0 and then not update it

**Comment by mahnerak** [comment]
Let's consider your model containing LSTM layer is defined like this:
```python
'''
Model gets temporal data and tries to predict one of labels of 10 classes.
'''
x = Input(shape=(200, 50))
h1 = LSTM(units=20, name='layer_to_freeze') (x)
h2 = LSTM(units=20) (x)
h = Concatenate() ([h1, h2])
y = Dense(units=10, activation='softmax')

model = Model(inputs=x, outputs=y)
```
Suppose you train the model after this. 
Then if you want to freeze the layer:
```python
layer_to_freeze = model.get_layer('layer_to_freeze') # Get the layer you want to freeze
layer_to_freeze.trainable = False # Force not to update layer params
old_weights = layer_to_freeze.get_weights() # Get old weights
new_weights = [0 * weight for weight in old_weights] # Zero-filled arrays with same shapes
layer_to_freeze.set_weights(new_weights) # Set zero-filled weights
```

**Comment by shaifugpt** [comment]
This will stop training of all connections associated with a particular layer. I want to stop training of some weight connections from each layer

**Comment by mahnerak** [comment]
Do you want to freeze only part of weight matrix?
I think it's impossible with Keras LSTM layer. You have to implement (or at least inherit base LSTM and extend the layer) to support weight freezing like that.

**Comment by shaifugpt** [comment]
Ok. Can you give an idea on how can it be performed. Will extending the layer be sufficient to freeze specific weights of all layers

**Comment by mahnerak** [comment]
Better way to freeze part of weights matrices:
1. Make element-wise equivalent of MaxNorm Constraint
(https://github.com/fchollet/keras/blob/master/keras/constraints.py#L17)
2. Define the model
3. Train the model
4. Decide which elements you want to freeze. Make `np.array` with zeros for freezing elements and `np.inf` for others and name it `constraint_mask`.
5. Get layer, set constraints on weights
```python
kernel = model.get_layer('layer_to_freeze').kernel
kernel.constraint = ElemwiseMaxNorm(constraint_mask, axis=())
```
6. Define new optimizer and compile the model again
7. Train the model
8. ...

**Comment by shaifugpt** [comment]
Ok .. I will try and post result

**Comment by shaifugpt** [comment]
Will it be correct to keep np.inf for others? Instead of infinity, can we prefer using a number like 2 or 3 to constrain the norm of weights

**Comment by shaifugpt** [comment]
@mahnerak why there is a need to use np.inf. Can we use 1 or any other number in its place

**Comment by mahnerak** [comment]
@shaifugpt Exactly. It strictly depends on your custom constraint whether using np.inf is needed or not. So yes, you can define a cust constraint which gets 0/1 valued np array as a mask which shows if the coresponding element is going to be clipped (zeroed).

**Comment by shaifugpt** [comment]
I created a class as:

      class clipweights(Constraint):
         def __call__(self, w):
            return K.clip(w,-10,0)

model.get_layer('layer_to_clip').kernel.constraint = clipweights()

but still after recompiling and model building, the max weight is greater than 0. Why?

**Comment by mahnerak** [comment]
As far as I know, the constraint is applied by optimizer after training step.
So try to do at least one step of training then check the weight values.

**Comment by shaifugpt** [comment]
I have recompiled and refit the model using model.fit and then the weights are checked

**Comment by mahnerak** [comment]
What version of Keras do you use?

**Comment by shaifugpt** [comment]
@mahnerak keras-2.0.5

**Comment by mahnerak** [comment]
Looks like this approach is supported in latest version of Keras (not released yet, 2.0.7).
I forgot to mention it (sorry about that).
You can try uninstalling your current Keras version, then installing the latest version from git:
```sh
pip install git+https://github.com/fchollet/keras
```
(It will show 2.0.6 is installed, but it's not)

**Comment by shaifugpt** [comment]
@mahnerak Still no sucess. The weights are still not clipped. It is shown but I guess not applied

model.get_layer('layer_to_freeze').kernel_constraint: prints the constraint name as
<__main__.clipweights object at 0x7f86c03a3eb8>

**Comment by mahnerak** [comment]
Please look at my code, it works as expected!

```python
from keras import backend as keras
from keras.models import Model
from keras.layers import Input, LSTM, Concatenate, Dense
from keras.constraints import NonNeg

'''
Model gets temporal data and tries to predict one of labels of 10 classes.
'''

x = Input(shape=(200, 50))
h1 = LSTM(units=20, name='layer_to_freeze') (x)
h2 = LSTM(units=20) (x)
h = Concatenate() ([h1, h2])
y = Dense(units=10, activation='softmax') (h)

model = Model(inputs=x, outputs=y)

import numpy as np
x0 = np.random.randn(150, 200, 50)
y0 = np.random.randn(150, 10)

# here you compile and train

model.compile(loss='categorical_crossentropy',
              optimizer='SGD',
              metrics=['accuracy'])
model.fit(x0, y0)

# then set the constraint

model.get_layer('layer_to_freeze').kernel.constraint = NonNeg()

# compile (and train at least one step) again

model.compile(loss='categorical_crossentropy',
              optimizer='SGD',
              metrics=['accuracy'])
model.fit(x0, y0)

# now, lets examine our weights

kernel, bias, rec = model.get_layer('layer_to_freeze').get_weights()

print('kernel min:', kernel.min())
print('bias min:', bias.min())
print('rec min:', rec.min())

```

It will print results like this:
```
(u'kernel min:', -0.0)
(u'bias min:', -0.38109347)
(u'rec min:', -0.004047065)
```

**Comment by shaifugpt** [comment]
@mahnerak Yes it seems to be working in your code. I will try it. Thanks
